### PR TITLE
image: Fix context hash of symlinks in subdirs

### LINKED ISSUE #541: Version 4.x: unable to hash build context ... could not evaluate symlink
### What happened?

```
Diagnostics:
  docker:index:Image (other-project):
    error: unable to hash build context: error while hashing "project/.eslintrc.json": could not evaluate symlink atproject/.eslintrc.json: lstat project: no such file or directory
```

### Expected Behavior

Docker image could be built as it was built in 3.x

### Steps to reproduce

Add a symlink within the code of context

### Output of `pulumi about`

```
CLI          
Version      3.57.1
Go Version   go1.20.1
Go Compiler  gc

Plugins
NAME          VERSION
digitalocean  4.18.0
docker        4.0.0
kubernetes    3.24.1
nodejs        unknown
postgresql    3.6.0

Host     
OS       fedora
Version  37
Arch     x86_64

This project is written in nodejs: executable='/run/user/1000/fnm_multishells/6117_1678693753068/bin/node' version='v18.12.1'

.........

Dependencies:
NAME                  VERSION
@pulumi/kubernetes    3.24.1
@pulumi/postgresql    3.6.0
@pulumi/pulumi        3.57.1
@types/node           18.15.1
prettier              2.8.4
typescript            4.9.5
@pulumi/digitalocean  4.18.0
@pulumi/docker        4.0.0

Pulumi locates its logs in /tmp by default
```

### Additional context

This symlink references to another symlink (i need to have it like this)

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by AaronFriel**
Fixes #541.

Evaluates symlinks using the full path to the file, not the incorrect relative path to the context dir. The provider's current working directory and the context dir are likely to be different. Regular files are treated similarly:

https://github.com/pulumi/pulumi-docker/blob/8cd8688eaee94c588d65f18862d95a1b0ff3f247/provider/provider.go#L455-L460

### CODE REVIEW

**Comment by mikhailshilkov** [comment]
Is there a way to add a test for this?

**Comment by AaronFriel** [comment]
@mikhailshilkov a unit test of the context hasher makes sense, I'll add that.

File: provider/provider_test.go [hunk 1]
```diff
@@ -106,3 +108,38 @@ func TestDiffUpdates(t *testing.T) {
 	})
 
 }
+
+func TestHashIgnoresFile(t *testing.T) {
+	baseResult, err := hashContext("./testdata/ignores/basedir", "./Dockerfile")
```
Comment (guineveresaenger) [inline provider/provider_test.go:113]: Can we use `Dockerfile` without the `./` in a few places? The defaults do not include the dot-slash.

**Review by guineveresaenger — APPROVED** [review]
nit, but LGTM.

**Pull request merged**